Toolkit/ArcGISToolkit/JobManager.swift

```diff
@@ -55,80 +55,57 @@ public class JobManager: NSObject {
     
     private var kvoContext = 0
     
-    deinit {
-        jobs.forEach { unObserveJobStatus(job: $0) }
-    }
-    
-    public private(set) var keyedJobs = [String: AGSJob](){
-        didSet{
-            self.updateJobsArray()
-            saveJobsToUserDefaults()
+    public private(set) var keyedJobs = [String: AGSJob]() {
+        willSet {
+            // Need `self` because of a Swift bug.
+            self.keyedJobs.values.forEach { unObserveJobStatus(job: $0) }
+        }
+        didSet {
+            keyedJobs.values.forEach { observeJobStatus(job: $0) }
+
+            // If there was a change, then re-store the serialized jobs in UserDefaults
+            if keyedJobs != oldValue {
+                saveJobsToUserDefaults()
+            }
         }
     }
-    public private(set) var jobs = [AGSJob]()
-    private func updateJobsArray(){
-        
-        // when our jobs array changes we need to observe the jobs' status
-        // that we aren't currently observing. The best way to do that is to
-        // just unObserve all, then re-observe all job status events
-        
-        // so first un-observe all current jobs
-        jobs.forEach { unObserveJobStatus(job: $0) }
-        
-        // set new jobs array
-        jobs = keyedJobs.map{ $0.1 }
-        
-        // now observe all jobs
-        jobs.forEach { observeJobStatus(job: $0) }
+
+    public var jobs: [AGSJob] {
+        return Array(keyedJobs.values)
     }
     
-    private func toJSON() -> JSONDictionary{
-        var d = [String: Any]()
-        for (jobID, job) in self.keyedJobs{
-            if let json = try? job.toJSON(){
-                d[jobID] = json
-            }
-        }
-        return d
+    private var jobsDefaultsKey: String {
+        return "com.esri.arcgis.runtime.toolkit.jobManager.\(jobManagerID).jobs"
     }
     
     /// Create a JobManager with an ID.
-    public required init(jobManagerID: String){
+    public required init(jobManagerID: String) {
         self.jobManagerID = jobManagerID
         super.init()
-        if let d = UserDefaults.standard.dictionary(forKey: self.jobsDefaultsKey){
-            suppressSaveToUserDefaults = true
-            self.instantiateStateFromJSON(json: d)
-            suppressSaveToUserDefaults = false
-        }
+        loadJobsFromUserDefaults()
     }
     
-    private func instantiateStateFromJSON(json: JSONDictionary){
-        for (jobID, value) in json{
-            if let jobJSON = value as? JSONDictionary{
-                if let job = (try? AGSJob.fromJSON(jobJSON)) as? AGSJob{
-                    self.keyedJobs[jobID] = job
-                }
-            }
-        }
+    deinit {
+        keyedJobs = [:]
```

Comment:
Is this intended to stop observing the the jobs? As far as I understand, property observers are not called when the property is set from inside `deinit`. However, I recall iOS 11 / macOS 10.13 introducing some changes to KVO where remaining observers after the object is deallocated will be cleaned up by KVO. That may be happening here.

That being said, I just had another idea. If you use the newer Swift closure-based KVO API, that could simplify things somewhat. [Here](https://github.com/Esri/arcgis-runtime-toolkit-ios/commit/e0f039ae083dccd1ba8bcb5b0f581787eed4160c) is what that would look like.

Reply:
I just realized deinit wasn't getting called in my tests, so this does a very bad thing. I'll back this change out.

Reply:
And yes, KVO is resilient to not having observers removed now, but still not a bad practice to explicitly do it.